### DISCUSSION TITLE
Apache bench seems to have issues with uWebSockets.js

### DISCUSSION THREAD

**Discussion by jbenguira** [uNetworking/uWebSockets.js#114] [Q&A] [answered]
I just tried to run the hello world sample (untouched), it seems to be running fine when I try it in my browser but when I try with ab (apache bench) I get this strange output (and 12 sec time per request)

Server Port:            80

Document Path:          /
Document Length:        12 bytes

Concurrency Level:      128
Time taken for tests:   48.210 seconds
Complete requests:      512
Failed requests:        0
Keep-Alive requests:    0
Total transferred:      45440 bytes
HTML transferred:       7680 bytes
Requests per second:    10.62 [#/sec] (mean)
Time per request:       12052.471 [ms] (mean)
Time per request:       94.160 [ms] (mean, across all concurrent requests)
Transfer rate:          0.92 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        1    4   1.8      4       8
Processing:  8017 11013 1728.4  12008   12023
Waiting:        1    4   1.5      4       8
Total:       8018 11017 1729.2  12012   12024

Percentage of the requests served within a certain time (ms)
  50%  12012
  66%  12015
  75%  12017
  80%  12018
  90%  12020
  95%  12022
  98%  12023
  99%  12024
 100%  12024 (longest request)

Any idea why and how to fix this? :)
Thanks a lot

**Comment by slavaGanzin** [comment]
Will switching to https://github.com/loadimpact/k6 *fix this*?

**Comment by ghost** [comment]
Sadly projects like AB still see usage today even though it runs the now 25 year old http 1.0 protocol, which no modern web browser or client talks. We only support / acknowledge protocols that see actual real world usage today.

By refusing to fix this thing, it's easy to spot when you do user http 1.0, such as is the case with nginx proxying by default. Fix by swapping to 1.1

**Comment by ghost** [comment, marked as answer]
Wrk is probably the most well known alternative to ab

**Comment by jbenguira** [comment]
Thanks a lot Alex, I was not aware of Wrk & the situation with ab

I've just tested with Wrk on my test VM (Hetzner CX21 - 2 Skylake vcores - 4gb ram) for a simple hello world with Fastify VS uWebSocket.js

**Fastify running in cluster mode with 2 instances**
root@ubuntu-4gb-nbg1-1:~/wrk# wrk -t2 -c16 -d7s http://127.0.0.1:3000/
Running 7s test @ http://127.0.0.1:3000/
  2 threads and 16 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   338.60us  259.61us  10.75ms   98.36%
    Req/Sec    24.73k   536.96    25.66k    74.29%
  344542 requests in 7.01s, 53.89MB read
Requests/sec:  49169.84
Transfer/sec:      7.69MB

**uWebSocket.js running with just node (is crashing in cluster mode because the port is already in use)**
root@ubuntu-4gb-nbg1-1:~/wrk# wrk -t2 -c16 -d7s http://127.0.0.1:80/
Running 7s test @ http://127.0.0.1:80/
  2 threads and 16 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.08ms    2.22ms  16.08ms   85.56%
    Req/Sec    78.60k     7.71k  148.41k    88.49%
  1101385 requests in 7.09s, 74.58MB read
Requests/sec: 155259.24
Transfer/sec:     10.51MB

3 times better than Fastify, Wow!

**Comment by jbenguira** [comment]
While I'm testing with Wrk,

I have tested a small real world use case: an auth api calling a local redis server to store/fetch the data

Fastify (mono process with node): 30 625 Requests/sec
uWebSocket.js (mono process with node): 23 535 Requests/sec

I've re-runned the test several time because it was contradictory with the hello world test, and I always get the same results. Not sure why.

**Comment by ghost** [comment]
I can't help you with that here, you can always send me a mail for professional help

**Comment by s0kil** [comment]
@jbenguira Could you show the code you have for uWS?

**Comment by jbenguira** [comment]
> @jbenguira Could you show the code you have for uWS?

I've just rerunned now with the latest version, unfortunately I don't have access to the same hardware than 1 month ago.

Also now both version produce similar results: ~12K RPS


/////////////////// uWebSockets.js ////////////////////
const uWS = require('uWebSockets.js');
const port = 9001;

var redis = require("redis");
var client = redis.createClient();
client.on("error", function (err) {
    console.log("Error " + err);
});
const app = uWS.App().get('/*', async (res, req) => {
  /* Can't return or yield from here without responding or attaching an abort handler */
  res.onAborted(() => { res.aborted = true; });
  client.incr("counter"); //increment a counter in redis for each api call
  //read the latest counter value and return it
  client.get("counter", function (err, reply) {
      if (!res.aborted) { res.end( reply ); }
  });
}).listen(port, (token) => {
  if (token) { console.log('Listening to port ' + port); }
  else { console.log('Failed to listen to port ' + port); }
});


-----------------------

/////////////////// Fastify ////////////////////
var redis = require("redis");
var client = redis.createClient();
client.on("error", function (err) {
    console.log("Error " + err);
});
const fs = require('fs'); const path = require('path');
const fastify = require('fastify')({logger: true, http2: false});

fastify.get('/', (request, reply) => {
    //increment a counter in redis for each api call
    client.incr("counter");
    //read the latest counter value and return it
    client.get("counter", function (err, redisReply) {
        reply.send(redisReply);
    });
})
fastify.listen(9001, "0.0.0.0", err => {
  if (err) throw err
  console.log(`server listening on ${fastify.server.address().port}`)
})
